crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,105 @@
+use std::cmp;
+use std::io::{Write};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+use virtio_queue::{DescriptorChain};
+use vm_memory::{Bytes, GuestMemory};
+
+/// Console device resize feature.
+pub const VIRTIO_CONSOLE_F_SIZE: u64 = 0;
+
+/// Console device multiport feature.
+pub const VIRTIO_CONSOLE_F_MULTIPORT: u64 = 1;
+
+/// Console device emergency write feature.
+pub const VIRTIO_CONSOLE_F_EMERG_WRITE: u64 = 2;
+
+
+/// Console struct
+pub struct Console {
+    /// Data to send to the driver via receiveq
+    input_buffer: Arc<Mutex<Vec<u8>>>,
+
+    // /// Driver acked features
+    // features: u64,
+}
+
+impl Console {
+    /// Create new console object
+    pub fn new() -> Self{
+        Console {
+            input_buffer: Arc::new(Mutex::new(Vec::new())),
+            // features,
+        }
+    }
+
+    // fn has_feature(&self, feature_pos: u64) -> bool {
+    //     (self.features & (1u64 << feature_pos)) != 0
+    // }
+
+    /// Add data to the input buffer
+    pub fn enqueue_data(&self, data: &mut Vec<u8>) {
+        let mut input_buffer = self.input_buffer.lock().unwrap();
+        input_buffer.append(data);
+    }
+
+    /// Check if input buffer is empty
+    pub fn is_input_buffer_empty(&self) -> bool {
+        let input_buffer = self.input_buffer.lock().unwrap();
+        input_buffer.is_empty()
+    }
+
+    /// Process transmitq chain
+    pub fn process_transmitq_chain<M, T>(&self, desc_chain: &mut DescriptorChain<M>, output: &mut T) -> usize
```

Comment:
I thought it would be more flexible to have the `output` sink given as a parameter instead of being part of the `Console`. This way we can change it on the fly, without having to modify or reinitialize the `Console`. I am not sure how useful this is, I could make it part of the `Console` to have it more unified.

Reply:
Why do we need to change the output on the fly?

Reply:
For example, I was thinking you could redirect it to a file in case of a large transfer and when that is done you could go back to `stdout`.

Reply:
This can be a future improvement if we end up needing it. For now I think it's better to also take care of the output similarly to the input.